File: Arrays/Leetcode/Medium/GroupAnagrams/GroupAnagrams.cpp
// https://leetcode.com/problems/group-anagrams/
#include<bits/stdc++.h>
using namespace std;

string hashKey(string &a)
{
    vector<int> hashMap(26, 0);

    for (int i = 0; i < a.length(); i++)
    {
        hashMap[a[i] - 'a'] += 1;
    }
    string key = "";
    for (int &i : hashMap)
    {
        key += to_string(i) + ",";
    }
    return key;
}
vector<vector<string>> groupAnagrams(vector<string> &strs)
{
    unordered_map<string, vector<string>> mp;
    for (string &str : strs)
    {
        mp[hashKey(str)].emplace_back(str);
    }

    vector<vector<string>> ans;
    for (auto &values : mp)
    {
        ans.emplace_back(values.second);
    }
    return ans;
}

File: Arrays/Leetcode/Medium/GroupAnagrams/GroupAnagrams.java
// https://leetcode.com/problems/group-anagrams/
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GroupAnagrams {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();

        for(String str: strs) {
            int hashMap[] = new int[26];
            for(int i=0; i<str.length(); i++) {
                hashMap[str.charAt(i) - 'a']++;
            }
            String key = Arrays.toString(hashMap);
            map.computeIfAbsent(key, k -> new ArrayList<>()).add(str);
        }

        return new ArrayList<>(map.values());
    }
}
